lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -225,56 +225,40 @@ sub save_results {
 
-# Collect text output into JSON. save_results() is called too early for this.
+# incorporate textual step data into details JSON
+# note: Can not be called from save_results() because the upload must have already been concluded.
 sub finalize_results {
     my ($self) = @_;
 
-    my $dir = $self->job->result_dir();
-    return 0 unless $dir;
-    my $file    = path($dir, "details-" . $self->name . ".json");
-    my $tmpfile = $file->sibling($file->basename . '.tmp');
-    return 0 unless -e $file;
-    my $errors = 0;
-    my $results;
-
-    eval {
-        $results = decode_json($file->slurp);
-        $results = {details => $results} if ref($results) eq 'ARRAY';
-
-        for my $step (@{$results->{details}}) {
-            next unless $step->{text};
-            my $txtfile = path($dir, $step->{text});
-            next unless -e $txtfile;
-            $step->{text_data} = decode('UTF-8', $txtfile->slurp);
-        }
+    # locate details JSON; skip if not present
+    my $dir = $self->job->result_dir;
+    return undef unless $dir;
+    my $name = $self->name;
+    my $file = path($dir, "details-$name.json");
+    return undef unless -e $file;
 
-        $tmpfile->spurt(encode_json($results));
-
-        rename $tmpfile->to_string, $file->to_string
-          or die "finalize_results(" . $self->job->id . "): Cannot update details-" . $self->name . ".json. $!";
-    };
-
-    if (my $err = $@) {
-        log_error $err;
-        $errors = 1;
+    # read details; skip if none present
+    my $results = decode_json($file->slurp);
+    my $details = ref $results eq 'HASH' ? $results->{details} : $results;
+    return undef unless ref $details eq 'ARRAY' && @$details;
+
+    # incorporate textual step data into details
+    for my $step (@$details) {
+        next unless my $text = $step->{text};
+        my $txtfile = path($dir, $text);
+        next unless -e $txtfile;
+        $step->{text_data} = decode('UTF-8', $txtfile->slurp);
     }
 
-    eval { $tmpfile->remove if -e $tmpfile; };
-    return !$errors;
-}
-
-# Delete result files which were merged into JSON details file.
-sub cleanup_results {
-    my ($self) = @_;
-
-    my $dir = $self->job->result_dir();
-    return unless $dir;
-    my $file = path($dir, "details-" . $self->name . ".json");
-    return unless -e $file;
-    my $results = decode_json($file->slurp);
+    # replace file contents on disk using a temp file to preserve old file if something goes wrong
+    my $new_file_contents = encode_json($results);
+    my $tmpfile           = tempfile(DIR => $file->dirname);
+    $tmpfile->spurt($new_file_contents);
+    $tmpfile->move_to($file);
 
-    for my $step (@{$results->{details}}) {
-        next if !$step->{text} || !defined($step->{text_data});
+    # cleanup incorporated files
+    for my $step (@$details) {
+        next unless $step->{text} && defined $step->{text_data};
         my $textfile = path($dir, $step->{text});
         $textfile->remove if -e $textfile;
```

Comment:
If you collect and delete the text files in a single step, concurrent WebUI requests that have loaded the old JSON file will show bogus errors to the user.

Reply:
That's true. However, cleaning up in a separate loop or even a separate Minion job doesn't really fix that race condition. It just acts like a "sleep". Maybe that is better handled when the possibly old JSON file is being processed.

Reply:
If you set the delay between the two minion jobs higher than Apache request timeout, the race condition is fixed. But in practice, the WebUI request only needs to finish processing one JSON file between the two minion jobs. The next JSON file will already include all the available text data.

Reply:
I'm not even using Apache here any this should likely work independently from the Apache request timeout. I guess it is the simplest to let `OpenQA::Schema::Result::JobModules::results` handle this situation by simply trying to read the details file one more time in case it has been "extended". This way we don't have to introduce complicated locking or rely on certain timeouts.

The only problem is that this situation is hard to test.